service/lib/dinstaller/dbus/manager.rb

```diff
@@ -20,68 +20,79 @@
 
 require "dbus"
-require "dinstaller/manager"
 
 module DInstaller
   module DBus
-    # YaST D-Bus object (/org/opensuse/DInstaller/Manager1)
-    #
-    # @see https://rubygems.org/gems/ruby-dbus
+    # D-Bus object to manage the installation process
     class Manager < ::DBus::Object
       PATH = "/org/opensuse/DInstaller/Manager1"
       private_constant :PATH
 
       MANAGER_INTERFACE = "org.opensuse.DInstaller.Manager1"
       private_constant :MANAGER_INTERFACE
 
-      attr_reader :logger
-
+      # Constructor
+      #
+      # @param backend [DInstaller::Manager] Installation manager
       # @param logger [Logger]
-      def initialize(logger)
+      def initialize(backend, logger)
         @logger = logger
-        @backend = ::DInstaller::Manager.instance.tap(&:probe)
-        @backend.progress.add_on_change_callback do
-          PropertiesChanged(MANAGER_INTERFACE, { "Progress" => progress }, [])
-        end
-        @backend.add_status_callback do
-          PropertiesChanged(MANAGER_INTERFACE, { "Status" => @backend.status.id }, [])
-        end
+        @backend = backend
+
+        add_status_callback
+
+        add_progress_callback
 
         super(PATH)
       end
 
       dbus_interface MANAGER_INTERFACE do
-        dbus_method :Probe, "" do
-          @backend.probe
-        end
+        dbus_method(:Probe, "") { backend.probe }
 
-        dbus_method :Commit, "" do
-          @backend.install
-        end
+        dbus_method(:Commit, "") { backend.install }
 
-        # Enum list for statuses. Possible values:
-        # 0 : error ( it can be read from progress message )
-        # 1 : probing
-        # 2 : probed
-        # 3 : installing
-        # 4 : installed
+        # Current status
+        #
+        # Possible values:
+        #   0 : error
+        #   1 : probing
+        #   2 : probed
+        #   3 : installing
+        #   4 : installed
         dbus_reader :status, "u"
 
+        # Error messages when the status is 0
+        dbus_reader :error_messages, "as"
+
         # Progress has struct with values:
-        # s message
-        # t total major steps to do
-        # t current major step (0-based)
-        # t total minor steps. Can be zero which means no minor steps
-        # t current minor step
+        #   s message
+        #   t total major steps to do
+        #   t current major step (0-based)
+        #   t total minor steps. Can be zero which means no minor steps
+        #   t current minor step
         dbus_reader :progress, "(stttt)"
       end
 
+      # Id of the current status
+      #
+      # @return [Integer]
       def status
-        @backend.status.id
+        backend.status_manager.status.id
+      end
+
+      # Error messages from the error status
+      #
+      # @return [Array<String>]
+      def error_messages
+        return [] unless backend.status_manager.error?
+
+        backend.status_manager.status.messages
```

Comment:
Right now it only includes errors from the Error status. And Error status is storing only the error message from the lastest exception. But in the API I have considered the possibility of returning more than one error message, just in case we need to start reporting serveral errors.